### PR TITLE
[cloud] Add support for updating IAM role with ec2_instance module

### PR DESCRIPTION
**Pull request by ryansb**
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
IAM instance profiles can't be switched via the `modify_instance_attribute` API call. This pull request adds checks on already-existing instances for the IAM instance profile ARN and name, then updates the profile association. 

<!---
If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
ec2_instance

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```

```


##### ADDITIONAL INFORMATION
<!---
Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful.
  -->


Before:

```
        "instances": [                                                                                                                                                                                              
            {                                                                                                                                                                                                       
                "ami_launch_index": 0,                                                                                                                                                                              
                "architecture": "x86_64",                                                                                                                                                                           
                "block_device_mappings": [                                                                                                                                                                          
                    {                                                                                                                                                                                               
                        "device_name": "/dev/sda1",                                                                                                                                                                 
                        "ebs": {                                                                                                                                                                                    
                            "attach_time": "2018-04-13T19:52:22+00:00",                                                                                                                                             
                            "delete_on_termination": false,                                                                                                                                                         
                            "status": "attached",                                                                                                                                                                   
                            "volume_id": "vol-0f3d246ba2ba774c9"                                                                                                                                                    
                        }                                                                                                                                                                                           
                    }                                                                                                                                                                                               
                ],                                                                                                                                                                                                  
                "client_token": "a0ef26cc419a4eda822c7ac8dd0d4ec9",                                                                                                                                                 
                "ebs_optimized": false,                                                                                                                                                                             
                "ena_support": true,                                                                                                                                                                                
                "hypervisor": "xen",                                                                                                                                                                                
                "image_id": "ami-0ebdd976",                                                                                                                                                                         
                "instance_id": "i-088a1fbf6d42ad068",                                                                                                                                                               
                "instance_type": "t2.micro",                                                                                                                                                                        
                "key_name": "slscode_hornet_2018",                                                                                                                                                                  
                "launch_time": "2018-04-13T19:52:21+00:00",    
```
<!--- Paste verbatim command output below, e.g. before and after your change -->
After:

```
    "instances": [                                                                                                                                                                                                  
        {                                                                                                                                                                                                           
            "ami_launch_index": 0,                                                                                                                                                                                  
            "architecture": "x86_64",                                                                                                                                                                               
            "block_device_mappings": [                                                                                                                                                                              
                {                                                                                                                                                                                                   
                    "device_name": "/dev/sda1",                                                                                                                                                                     
                    "ebs": {                                                                                                                                                                                        
                        "attach_time": "2018-04-13T19:52:22+00:00",                                                                                                                                                 
                        "delete_on_termination": false,                                                                                                                                                             
                        "status": "attached",                                                                                                                                                                       
                        "volume_id": "vol-0f3d246ba2ba774c9"                                                                                                                                                        
                    }                                                                                                                                                                                               
                }                                                                                                                                                                                                   
            ],                                                                                                                                                                                                      
            "client_token": "a0ef26cc419a4eda822c7ac8dd0d4ec9",                                                                                                                                                     
            "ebs_optimized": false,                                                                                                                                                                                 
            "ena_support": true,                                                                                                                                                                                    
            "hypervisor": "xen",                                                                                                                                                                                    
            "iam_instance_profile": {                                                                                                                                                                               
                "arn": "arn:aws:iam::509803855674:instance-profile/otherTestRole",                                                                                                                                  
                "id": "AIPAJBBIY7IVIPOJ2SSDW"                                                                                                                                                                       
            },                                                                                                                                                                                                      
            "image_id": "ami-0ebdd976",                                                                                                                                                                             
            "instance_id": "i-088a1fbf6d42ad068",  
```

### CODE REVIEW

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 1]
```diff
@@ -737,6 +737,47 @@ def build_volume_spec(params):
...
+    if instance_profile_setting and desired_profile_name:
+        if desired_profile_name in (instance_profile_setting.get('Name'), instance_profile_setting.get('Arn')):
+            # great, the profile we asked for is what's there
+            return False
+        else:
+            desired_arn = determine_iam_role(desired_profile_name)
+            if instance_profile_setting.get('Arn') == desired_arn:
+                return False
+        # update association
+        ec2 = module.client('ec2')
```
Comment (s-hertel) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py:751]: I keep forgetting module is global :-)

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 1 above)
Comment (willthames) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py:751]: Is it? That seems wrong to me. Is it global in every module?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py [hunk 2]
```diff
@@ -1292,9 +1333,9 @@ def pretty_instance(i):
 def determine_iam_role(name_or_arn):
     if re.match(r'^arn:aws:iam::\d+:instance-profile/[\w+=/,.@-]+$', name_or_arn):
         return name_or_arn
-    iam = module.client('iam')
+    iam = module.client('iam', retry_decorator=AWSRetry.jittered_backoff())
     try:
-        role = iam.get_instance_profile(InstanceProfileName=name_or_arn)
+        role = iam.get_instance_profile(InstanceProfileName=name_or_arn, aws_retry=True)
```
Comment (willthames) [inline lib/ansible/modules/cloud/amazon/ec2_instance.py:1338]: We should update the [GUIDELINES](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/GUIDELINES.md#api-throttling-and-pagination) if we're happy enough to use it now (it's so much nicer than creating `_backoff` methods with `@AWSRetry` decorators)

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py:1338]: I'm quite happy with it (so nice to use!) but s-hertel brought up an interesting point about it for library/common functions. In module_utils files, we probably can't expect the caller to always have a retry decorator turned on for whatever client they pass in, so I think for utils we still have to use the decorators directly for now. What do you think?

File: lib/ansible/modules/cloud/amazon/ec2_instance.py (hunk 1 above)
Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/ec2_instance.py:751]: It's not global in every module, but it is global in this module https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_instance.py#L1409

**Pull request merged**